feat(monitor): Support truncation of arguments for logging

abwesend890 opened the pull request:
This request includes 
- a utility mapping function, 
- logic to print fact-arguments with the WARN setting, 
- indicate when an action is used that is not supported, 
- print more details on parsing failures

abwesend890 committed:
feat(utility): map function, intended to function in a similar way as described here https://docs.python.org/3/library/functions.html#map

abwesend890 committed:
feat(monitor/logging): add arguments to facts in the warning describing final configurations

abwesend890 committed:
feat(monitor/logging): log error on unknown action used in specification file

abwesend890 committed:
feat(monitor/logging): improve error message

abwesend890 committed:
project structure: update .gitignore for IDE data

abwesend890 committed:
improvement(monitor/logging): Mention that the next listing belong to the next configuration

kevinmorio commented on monitor/configuration.go:
This change disallows any other events in the rewriting except for `RewriteEventName` and r`Eq`, etc. This breaks monitoring. What was the intention here?

abwesend890 replied:
commit bdad31a9ff62b53884ebb2f0857f3f7964c263d9 proposes a change that does not error but only warns the user about a Action being ignored by specmon.

This prevents the user expecting specmon to do something it actually does not implement

kevinmorio commented on monitor/monitor.go:
This helps debugging but it is hard to parse when arguments are large. It would be great to have a configurable parameter to set a maximum length for each parameter that is rendered then as `0xdeadcafe...{1024}` for example to indicate its full length.

Instead of using `utils.Map`, a string builder approach avoids internal allocation:

```
var b strings.Builder
for i, arg := range f.Args {
	if i > 0 {
		b.WriteString(", ")
	}
	b.WriteString(arg.String())
}
log.Warnf("  %s(%s)", f.Name, b.String())
```

abwesend890 replied:
cf46f107db3f69b8f5ede0e21087740a3461f54c implements a CLI argument `log-arg-truncate` which expects an integer.
CLI Argument may look like `--log-arg-truncate 10`

a fix for minor logging issues is in 7acfbba2810375400609da99acd98b49c792dc86

Output example:
```
time="2025-10-03T15:59:57+02:00" level=warning msg="  hashed(0xfe000030..., 0x9f782688...)"
time="2025-10-03T15:59:57+02:00" level=warning msg="  KDF_c_hs_traffic_key(0x9ea787c6...)"
time="2025-10-03T15:59:57+02:00" level=warning msg="  KDF_c_hs_traffic_iv(0x4ad1dba1...)"
time="2025-10-03T15:59:57+02:00" level=warning msg="  KDF_c_finished_secret(0x58ec92b1...)"
```


9a1c81705349aa583541f898d47bc2763be51d0b replaces the map function with a StringBuilder

abwesend890 replied:
9b3620c61f93f3931b1a659a5350c542a8cbae50 includes the actual length of the string which is truncated in curly backets 

example output:

```
time="2025-10-03T16:04:16+02:00" level=warning msg="  KDF_c_hs_traffic_iv(0x4ad1dba1...{26})"
time="2025-10-03T16:04:16+02:00" level=warning msg="  KDF_c_finished_secret(0x58ec92b1...{98})"
time="2025-10-03T16:04:16+02:00" level=warning msg="  ServerHandshakeSequenceNr(0x01)"
time="2025-10-03T16:04:16+02:00" level=warning msg="  DEBUG_PARSE(parse_hand...{71})"
time="2025-10-03T16:04:16+02:00" level=warning msg="  DEBUG_enc(0x08000015...{54}, 0x11f8096e...{54})"
```

kevinmorio commented on utils/utils.go:
With a direct loop , this function is not needed anymore.

abwesend890 replied:
the function was removed with 

cf46f107db3f69b8f5ede0e21087740a3461f54c

kevinmorio reviewed (changes requested):
Thanks for the PR @abwesend890!

I've added a few comments to the changes.

abwesend890 committed:
fix(monitor): to support spthy files for TAMARIN we can not error here. For developer convenience I propose any message that this action does not imply any logic within SPECMON.

abwesend890 committed:
feat(fact) logArgTruncate CLI-argument: specifies the maximum length of a fact's arguments before they are truncated in log output.
dependencies for that:
feat(monitor) settings map[string]interface{}: this may hold arbitrary configuration values set by root.go / shared.go to be passed to the monitor --- The map is used to store the logArgTruncate argument and pass it to the monitor.
feat(shared) handle int64 arguments: Previously, a int CLI argument was not honored
feat(utils) function to shorten a string to a given maxLength

abwesend890 committed:
improvement(fact): use more efficient string builder to print Fact arguments, commentary

abwesend890 committed:
fix(fact) LogArg: do not account for 0x as there may also be other data types...
fix(fact) LogArg: do not print "..." despite the string was not shortened, as this required the truncateString function to be more specific I moved it to the class file

abwesend890 committed:
feat(fact) LogArg: add max length in curly brackets

abwesend890 committed:
Merge branch 'develop' into feat/improve-logging-by-printing-fact-arguments-in-WARN

abwesend890 committed:
refactor(format): golangci-lint

abwesend890 committed:
Merge branch 'develop' into feat/improve-logging-by-printing-fact-arguments-in-WARN

abwesend890 commented:
you mentioned that you wanted to comment on something here but I did not spot anything?

abwesend890 committed:
Merge branch 'develop' into feat/improve-logging-by-printing-fact-arguments-in-WARN

kevinmorio commented on cmd/rewrite.go:
```diff
@@ -70,7 +70,15 @@ func (r *RewriteConfig) RunE(cmd *cobra.Command, args []string) error {
-	m, err = monitor.NewMonitor(decompRules)
+	// Retrieve User Settings
+	// factArgMaxLen specifies the maximum length of a fact's arguments before they are truncated in log output.
+	logArgTruncate, _ := cmd.Root().Flags().GetInt64("log-arg-truncate")
```
I think we can just call it `truncate-args`.

abwesend890 replied:
with 181b91d5d209293c7c9b7e53bb205dd3932dc333

kevinmorio commented on cmd/rewrite.go:
```diff
@@ -70,7 +70,15 @@ func (r *RewriteConfig) RunE(cmd *cobra.Command, args []string) error {
 	defer eventSource.Close()
 
-	m, err = monitor.NewMonitor(decompRules)
+	// Retrieve User Settings
```
Section comments are not needed here.

abwesend890 replied:
with 21958d6d8417ae0ef7261564ecf4bc381f572f73

kevinmorio commented on cmd/rewrite.go:
```diff
@@ -70,7 +70,15 @@ func (r *RewriteConfig) RunE(cmd *cobra.Command, args []string) error {
+
+	// Define User Settings for Monitor
+	settings := make(map[string]interface{})
+	settings["logArgTruncate"] = logArgTruncate
```
As above, just `"truncate-arg"`.

abwesend890 replied:
s.o.

kevinmorio commented on cmd/root.go:
```diff
@@ -48,6 +48,7 @@ type RootConfig struct {
 	SpecPath       string   `flag:"spec-path"        short:"s" desc:"specification path"`
 	CPUProfilePath string   `flag:"cpu-profile-path" short:"c" desc:"cpu profile path"`
 	MemProfilePath string   `flag:"mem-profile-path" short:"m" desc:"memory profile path"`
+	LogArgTruncate int64    `flag:"log-arg-truncate" desc:"Max length of logged arguments within facts. Default is -1 for no limit."`
```
As above, use `truncate-args`.

abwesend890 replied:
s.o.

kevinmorio commented on monitor/configuration.go:
```diff
@@ -266,6 +266,8 @@ func splitPair(t term.Term) (term.Term, term.Term) {
 func restrSatisfied(trace []*rule.Fact) error {
 	for _, t := range trace {
 		switch t.Name {
+		case RewriteEventName:
```
This doesn't belong to this function.

abwesend890 replied:
see below

kevinmorio commented on monitor/configuration.go:
```diff
@@ -280,6 +282,8 @@ func restrSatisfied(trace []*rule.Fact) error {
 			if t.Args[0].Equal(t.Args[1]) {
 				return fmt.Errorf("%w: %s", ErrRestrictionViolated, t)
 			}
+		default:
```
The function is on the hot-path. Hence, any unnecessary logging should be avoided so the default case should be removed.

abwesend890 replied:
I do not think this is unnecessary. 
If this case holds and the log is triggered, then there is a action specified which is not honored and the analyst should be informed about that. I do not care enough to continue forcing this in, but in theory, if the process is time critical, then the specification should not contain unnecessary Actions. And if the specification file does, the developer might have wrong expectations.

abwesend890 replied:
with 8a52480041d5844b5af9ee62f25caec9c788b43a

kevinmorio commented on monitor/monitor_test.go:
```diff
@@ -51,8 +53,16 @@ func TestMonitorMultipleFrFacts(t *testing.T) {
 
+	// Retrieve User Settings
+	// factArgMaxLen specifies the maximum length of a fact's arguments before they are truncated in log output.
+	logArgTruncate, _ := cmd.Root().Flags().GetInt64("log-arg-truncate")
```
As above, use `truncate-args`.

abwesend890 replied:
s.o.

kevinmorio commented on rule/fact.go:
```diff
@@ -250,3 +252,46 @@ func (f Facts) ExpandFacts(b *term.Binding) []*Fact {
 }
+
+// truncateString shortens a string to maxLen and appends '...' if the string was truncated.
+func truncateString(s string, maxLen int64) string {
```
This function should be moved to the `utils` package.

abwesend890 replied:
with 89c2906c9353b35884a9ad79910dcfc603d5085f

kevinmorio commented on monitor/monitor.go:
```diff
@@ -71,9 +71,12 @@ type Monitor struct {
 	stats *Stats
+
+	// settings hold user configurations to alter monitor behavior
+	settings map[string]interface{}
```
We should turn `settings` into a concrete structure type to get type safety.

abwesend890 replied:
with f4a4f2080b12f6ef655677b5c3344241e60882fc

kevinmorio reviewed (changes requested):
Thanks for the update! I left a few comments with minor suggestions.

abwesend890 committed:
refactor(monitor/logging) rename log-arg-truncate to truncate-args

abwesend890 committed:
fix(monitor/tests): add missing settings object as argument

abwesend890 committed:
refactor(rewrite/comments): remove section comments

abwesend890 committed:
revert(logging/hot-path): "this function is on the hot-path and should not contain unnecessary logging"
note: I do not think this is unnecessary.
If this case holds and the log is triggered, then the .spthy specifies an action which is not honored and the analyst should be informed about that. I do not care enough to continue forcing this in, but in theory, if the process is time critical, then the specification should not contain unnecessary actions. And if the specification file does, the developer might have wrong expectations.

abwesend890 committed:
refactor(utils/truncateString) move function truncateString to utils package and rename to TruncateString

abwesend890 committed:
improve(monitor/settings): turn settings into a concrete structure type to get type safety

abwesend890 committed:
fix(tests/settings): adapt passed settings to test runs

abwesend890 commented:
recent changes should include everything requested

abwesend890 committed:
refactor(gofumpt): import order

abwesend890 committed:
Merge branch 'develop' into feat/improve-logging-by-printing-fact-arguments-in-WARN

kevinmorio commented on data/settings.go:
```diff
@@ -0,0 +1,5 @@
+package data
+
+type Settings struct {
```
The `data` module is used for general purpose data structures. Since `Settings` is tied to the monitor, moving it to the `monitor` module would be better.

abwesend890 replied:
i tried this initially, but this leads to cyclic imports, which is why i moved it out

```
package github.com/specmon/specmon
	imports github.com/specmon/specmon/cmd from main.go
	imports github.com/specmon/specmon/monitor from monitor.go
	imports github.com/specmon/specmon/rule from configuration.go
	imports github.com/specmon/specmon/monitor from fact.go: import cycle not allowed

```

abwesend890 replied:
to be more specificy we may call it MonitorSettings, how do you think about that?

kevinmorio replied:
Thanks! I've solved the dependency issue by just passing `truncateArgs` to the `LogArgs` function., not the whole `Settings` structure.

kevinmorio reviewed (changes requested):
Thanks for the changes! One last nitpick: Since the `data` module is only used for general purpose data structures, `Settings` would be better placed in the `monitor` module next to the main `Monitor` type.

kevinmorio committed:
Move settings to monitor module
